jax/_src/callback.py

```diff
@@ -129,25 +135,50 @@ def callback_batching_rule(
     args,
     dims,
     *,
-    vectorized: bool,
+    vectorized: bool | None | DeprecatedArg,
+    vmap_method: str | None,
     result_avals: Sequence[core.ShapedArray],
     **kwargs: Any,
 ):
-  axis_size = next(a.shape[d] for a, d in zip(args, dims)
-                   if d is not batching.not_mapped)
+  if isinstance(vectorized, DeprecatedArg) and vmap_method is None:
+    warnings.warn(
+        f"The default behavior of {prim.name} under vmap will soon "
+        "change. Currently, the default behavior is to generate a sequential "
+        "vmap (i.e. a loop), but in the future the default will be to raise "
+        "an error. To keep the current default, set vmap_method='sequential'.",
+        DeprecationWarning, stacklevel=2)
```

Comment:
We probably need a higher stacklevel to make sure this displays at the call-site.

Reply:
Very good point! It's definitely tricky to set the right value here in general. For example, the following works (i.e. the warning points to the `jax.pure_callback` line) with `stacklevel=6`:

```python
import jax
import jax.numpy as jnp
import numpy as np

def f(x, **kwargs):
  return jax.pure_callback(np.sin, x, x, **kwargs)

jax.vmap(f)(jnp.arange(4.0))
```

But, I can't find any stack level that works for:

```python
jax.vmap(jax.jit(f))(jnp.arange(4.0))
```

In this second case, `stacklevel=6` points to an `eval_jaxpr`, but more importantly, `jax.pure_callback` doesn't appear anywhere in the stack.

The best I was able to come up with to cover these two cases was a scheme that combines the traceback filtering logic from `traceback_util` and `catch_warnings` to re-emit the warnings (see my latest commit). With these changes, we emitted warnings for these two examples are:

```
demo.py:6: DeprecationWarning: The default behavior of pure_callback under vmap will soon change. Currently, the default behavior is to generate a sequential vmap (i.e. a loop), but in the future the default will be to raise an error. To keep the current default, set vmap_method='sequential'.
  return jax.pure_callback(np.sin, x, x, **kwargs)
```

and

```
demo.py:9: DeprecationWarning: The default behavior of pure_callback under vmap will soon change. Currently, the default behavior is to generate a sequential vmap (i.e. a loop), but in the future the default will be to raise an error. To keep the current default, set vmap_method='sequential'.
  jax.vmap(jax.jit(f))(jnp.arange(4.0))
```

respectively.

WDYT? Is there a better way? Should we just choose a level (e.g. `6`) and be done with it?

Reply:
Yeah that's tricky... I'd probably lean against walking the stack in order to dynamically set the stacklevel. Maybe just set to 6, even though that may be an underestimate of the true stack level in the case of nested transformations.

Reply:
Yeah, I think that's probably better too. Reverted and set to 6.